mcfly/modelgen.py

```diff
@@ -115,630 +56,66 @@ def generate_models(x_shape,
 
+    # Add missing parameters from default
+    for key, value in defaults.items():
+        if key not in hyperparameter_ranges:
+            hyperparameter_ranges[key] = value
 
-    # Limit parameter space based on input
-    # -------------------------------------------------------------------------
-    if settings['IT_max_max_kernel_size'] > x_shape[1]:
-        print("Set maximum kernel size for InceptionTime models to number of timesteps.")
-        settings['IT_max_max_kernel_size'] = x_shape[1]
-    
     model_types_selected = []
-    model_types = settings['model_types']
 
-    for i in range(int(np.ceil(number_of_models/len(model_types)))):
+    if number_of_models%len(model_types) > 0:
+        number_of_models = \
+            int(np.ceil(number_of_models/len(model_types)))*len(model_types)
+        print("To have equal number of models for all types, number of models was changed to {}".format(number_of_models))
+
+    for i in range(number_of_models // len(model_types)):
         np.random.shuffle(model_types)
         model_types_selected.extend(model_types)
 
-    # Create list of Keras models
+    # Create list of Keras models and their hyperparameters
     # -------------------------------------------------------------------------
     models = []
-    for i in range(0, number_of_models):
-        current_model_type = model_types_selected[i]
-        generate_model = None
-        if current_model_type == 'CNN':
-            generate_model = generate_CNN_model  # generate_model is a function
-            hyperparameters = generate_CNN_hyperparameter_set(settings)
-        if current_model_type == 'DeepConvLSTM':
-            generate_model = generate_DeepConvLSTM_model
-            hyperparameters = generate_DeepConvLSTM_hyperparameter_set(settings)
-        if current_model_type == 'ResNet':
-            generate_model = generate_resnet_model
-            hyperparameters = generate_resnet_hyperparameter_set(settings)
-        if current_model_type == 'InceptionTime':
-            generate_model = generate_InceptionTime_model
-            hyperparameters = generate_InceptionTime_hyperparameter_set(settings)
-
-        models.append(
-            (generate_model(x_shape, 
-                            number_of_classes, 
-                            metrics=settings['metrics'], **hyperparameters),
-             hyperparameters, current_model_type))
+    for current_model_type in model_types_selected:
+        if current_model_type in default_models:
+            model_type = default_models[current_model_type](x_shape, number_of_classes,
+                                                            metrics, **hyperparameter_ranges)
+            hyperparameters = model_type.generate_hyperparameters()
+            model = model_type.create_model(**hyperparameters)
+            model_name = current_model_type
+        elif isinstance(current_model_type, str):
+            print("Unknown model name given.")
```

Comment:
better exception!

Reply:
Added an exception (and test to check if is indeed raised to ``test_modelgen.py``).